zjit/src/codegen.rs

```diff
@@ -605,6 +607,54 @@ fn gen_putspecialobject(asm: &mut Assembler, value_type: SpecialObjectType) -> O
     asm_ccall!(asm, rb_vm_get_special_object, ep_reg, Opnd::UImm(u64::from(value_type)))
 }
 
+fn gen_getspecial_symbol(asm: &mut Assembler, _key: u64, symbol_type: SpecialBackrefSymbol) -> Opnd {
+    // Fetch a "special" backref based on the symbol type
+
+    // call rb_backref_get()
+    asm_comment!(asm, "call rb_backref_get");
+    let backref = asm.ccall(rb_backref_get as *const u8, vec![]);
+
+    match symbol_type {
+        SpecialBackrefSymbol::LastMatch => {
+            asm_comment!(asm, "rb_reg_last_match");
+            asm.ccall(rb_reg_last_match as *const u8, vec![backref])
+        }
+        SpecialBackrefSymbol::PreMatch => {
+            asm_comment!(asm, "rb_reg_match_pre");
+            asm.ccall(rb_reg_match_pre as *const u8, vec![backref])
+        }
+        SpecialBackrefSymbol::PostMatch => {
+            asm_comment!(asm, "rb_reg_match_post");
+            asm.ccall(rb_reg_match_post as *const u8, vec![backref])
+        }
+        SpecialBackrefSymbol::LastGroup => {
+            asm_comment!(asm, "rb_reg_match_last");
+            asm.ccall(rb_reg_match_last as *const u8, vec![backref])
+        }
+    }
+}
+
+fn gen_getspecial_number(asm: &mut Assembler, nth: u64, state: &FrameState) -> Opnd {
+    // Fetch the N-th match from the last backref based on type shifted by 1
+
+    // call rb_backref_get()
+    asm_comment!(asm, "rb_backref_get");
+    let backref = asm.ccall(rb_backref_get as *const u8, vec![]);
+
+    // Save PC for GC
+    gen_save_pc(asm, state);
```

Comment:
We have a wrapper for this to clarify the intention by the function name.

```suggestion
    gen_prepare_call_with_gc(asm, state);
```